helsingborg-stad/helsingborg-io-sls-api
# Feat/creat viva case

## Explain the changes you’ve made
Adds support for looking up if a user are allowed to apply for an application period. If a user is allowed to apply a new initial case is created in AWS with VIVA information regarding the application period.

## Explain why these changes are made
VIVA controls if a user is allowed to apply for money on a specific period, since they have this functionality that we depend on we need to make sure that the data in our VIVA cases matches their underlaying system.

## Explain your solution
Each time a user completes a successful login in with BankID an event is passed to AWS eventBridge. The BankID complete event is what our new lambda(createVivaCase) listens to in order to start.

The lambda first does a check to se if the user have an open application period by retrieving a viva application status from our VIVA Adapter (located on prem in HBG Stad). The status returned from the request is then validated to se if a user is allowed to apply.

Once we know that a user is allowed to apply for a period we make a second request to the VADA Adapter to collect information about the period and it's related workflow and validate this information accordingly to the VIVA documentation for a recurring application period. 

After the validation towards VIVA the lambda creates an initial case with viva as a provider and populate the details attribute with viva specific information about the period and sets the status to openToApply.

## How to test the changes?
Setup and configuration
1. Checkout this branch.
2. Navigate to services/viva-ms and deploy the lambda in your personal environment.
```
// deploys the whole viva-ms service
$sls deploy

// deploy the createVivaCase lambda
$sls deploy function -f createVivaCase
```
3. Add new ssm paramter caseEnvs in AWS with the following value:
```
{
   "recurringFormId": "replace_this_with_a_exisiting_form_id",
}
```
Trigger lambda with event
1. In the AWS console go to the EventBridge service and navigate to [event buses](https://eu-north-1.console.aws.amazon.com/events/home?region=eu-north-1#/eventbuses) in the meny.
2. Select the only eventbus in the list that is presented (the name of the eventbus should be "default")
3. Once in the eventbus configuration panel you want to press the "Send Events" button in the up right corner.
4. Enter the following values in the input fields: Event Source: bankid.collect, Detail Type: BankIdCollectComplete, Text: {
  "user": {"personalNumber": "replace_this_value_with_a_personal_number"}
}

OBS make sure that the peronsalNumber that is a person that also exists in VADA in order to get a case to be created.

Check new case in dynamodd
When triggering the event for the first time and if it´s successfully there should be a new item in the dynamodb cases-table that looks like the following object:
```{
  "answers": [

  ],
  "createdAt": 1611758051108,
  "currentPosition": 1,
  "details": {
    "period": {
      "end": "2021-02-28",
      "start": "2021-02-01"
    },
    "workflowId": "459A8735121899A4C12586650047F916"
  },
  "expirationTime": 0,
  "formId": "0",
  "id": "ac48eedf-819f-4011-bbe8-da6cc2c8c5b4",
  "PK": "USER#195809262743",
  "provider": "VIVA",
  "SK": "USER#195809262743#CASE#ac48eedf-819f-4011-bbe8-da6cc2c8c5b4",
  "status": "openToApply",
  "updatedAt": 1611758051108
}```

Check for validation messages in Cloudwatch:
To see if the validation for the lambda works as intended you will need to watch the lambda logs found in AWS CloudWatch log groups. search for createVivaCase and then open the lambda in the list to get access to the logs.


## Was this feature tested in the following environments?
- [x] Personal AWS Environment.


## Anyhting else? (optional)
This lambda does not take in consideration if a user is present in the users table in AWS. This is due to that the collect response does not have this information. An improved event flow would be to listen for changes in the users table instead of the bankid collect event, but this behavior is not supported in our AWS configuration at the moment and can therefore not be used.
